### PR TITLE
Spi host flash tests

### PR DESCRIPTION
**Pull request by engdoreis**
This PR:
 - Refactor the DIF spi_host to allow the use of 4 lanes for the opcode section.
 - Add a new spi flash test for a gigadevice flash of 1Gb.

### CODE REVIEW

**Review by marnovandermaas — APPROVED** [review]
Looks good to me, but it would be good for someone with more experience writing software tests to have a look before merging.

**Comment by engdoreis** [comment]
I manually ran the nightly ci [job](https://dev.azure.com/lowrisc/opentitan/_build/results?buildId=120656&view=logs&j=ac8a02a7-2778-53b1-c82e-cd21a7d698f0&t=d325bfd8-fd6c-518b-a699-bb4f07334985) and it passed.

File: sw/device/lib/testing/spi_flash_testutils.h [hunk 1]
```diff
@@ -222,11 +246,10 @@ status_t spi_flash_testutils_erase_block64k(dif_spi_host_t *spih,
...
 OT_WARN_UNUSED_RESULT
-status_t spi_flash_testutils_program_op(dif_spi_host_t *spih, uint8_t opcode,
-                                        const void *payload, size_t length,
-                                        uint32_t address, bool addr_is_4b,
-                                        dif_spi_host_width_t data_width,
-                                        dif_spi_host_width_t addr_width);
+status_t spi_flash_testutils_program_op(
+    dif_spi_host_t *spih, uint8_t opcode, const void *payload, size_t length,
+    uint32_t address, bool addr_is_4b,
+    spi_flash_testutils_transaction_width_mode_t page_program_mode);
```
Comment (a-will) [inline sw/device/lib/testing/spi_flash_testutils.h:252]: nit: API comments need an update :)

File: sw/device/tests/spi_host_flash_test_impl.h [hunk 2]
```diff
@@ -108,8 +109,9 @@ status_t test_4bytes_address(dif_spi_host_t *spi);
  * @param addr_width The width of the transaction addressing section.
  * @return status_t containing either OK or an error.
  */
-status_t test_page_program_quad(dif_spi_host_t *spi, uint8_t opcode,
-                                uint8_t address_width);
+status_t test_page_program_quad(
+    dif_spi_host_t *spi, uint8_t opcode,
+    spi_flash_testutils_transaction_width_mode_t page_program_mode);
```
Comment (a-will) [inline sw/device/tests/spi_host_flash_test_impl.h:114]: nit: Update API doc comments.

**Review by a-will — APPROVED** [review]

**Pull request merged**